Add default HDMI port to fall back to when resetting

akirchhoff1 opened the pull request:
With a Pulse-Eight USB CEC adapter connected to HDMI2 (and nothing plugged into HDMI1) on a Samsung HG32NJ690FF or Samsung HG32NF693GF (haven't tested other Samsung models) with the TV set to always start up with TV as the source, if you power off the TV, then send the 'p 0 2' command in cec-client.exe, then send the 'as' command in cec-client.exe, you will see the TV power on to TV briefly, then switch to HDMI2, then about 10 seconds later, the TV switches to HDMI1 due to a physical address conflict that causes libcec to switch the TV to HDMI1 because it is resetting to its default HDMI port of HDMI1. We want the TV to show us HDMI2 when we send a command to libcec to turn on the TV and show us HDMI2.

opdenkamp commented:
can't you just start cec-client with the `-p` parameter to pass the correct port?

akirchhoff1 commented:
@opdenkamp That doesn't solve this problem. The issue here is that `CCECClient::ResetPhysicalAddress()` gets called in this situation by these Samsung TVs:
```
void CCECClient::ResetPhysicalAddress(void)
{
  LIB_CEC->AddLog(CEC_LOG_DEBUG, "resetting HDMI port and base device to defaults");
  SetHDMIPort(CECDEVICE_TV, CEC_DEFAULT_HDMI_PORT);
}
```

So this always forces the TV to switch to HDMI1 since that's what `CEC_DEFAULT_HDMI_PORT` is defined as.

That's why I want this section of code changed to:
```
void CCECClient::ResetPhysicalAddress(void)
{
  LIB_CEC->AddLog(CEC_LOG_DEBUG, "resetting HDMI port and base device to defaults");
  SetHDMIPort(CECDEVICE_TV, m_configuration.iDefaultHDMIPort);
}
```

akirchhoff1 commented:
@opdenkamp My goal is to power on the TV and show HDMI2 immediately (or as soon as possible) since that has the content that I'm trying to show. I don't know what would make that address "wrong" or how to otherwise correct it. I don't see this issue with various LG TV models, I've only ever had this issue with Samsung TVs. If it matters, the way I'm actually using this is with libcec as a DLL, but I was able to reproduce the problem using only `cec-client.exe` as described.

opdenkamp commented:
if you start `cec-client.exe -p 2` and then run your commands (without the `p 0 2` command), does it still reset the address in that case? if not: that'll be your fix. your app calling the dll has to pass that in the `libcec_configuration` at init.
if it still resets, then I'd like to see another log from this version, with the `-p 2` parameter.

thanks

akirchhoff1 commented:
@opdenkamp I'll try it out and get back to you. It's not likely to happen today though.

akirchhoff1 commented:
@opdenkamp Actually, I was able to locate the TV more easily than I thought and I still see the same result after `cec-client.exe -p 2` followed by `as`:
```
cec-client.exe -p 2
using HDMI port '2'
No device type given. Using 'recording device'
CEC Parser created - libCEC version 6.0.2
no serial port given. trying autodetect:
 path:     USB\VID_2548&PID_1002\v12
 com port: COM6

opening a connection to the CEC adapter...
DEBUG:   [               6]     Broadcast (F): osd name set to 'Broadcast'
DEBUG:   [              29]     connection opened, clearing any previous input and waiting for active transmissions to end before starting
DEBUG:   [              29]     communication thread started
DEBUG:   [              79]     usbcec: enabling controlled mode
NOTICE:  [             142]     connection opened
DEBUG:   [             142]     processor thread started
DEBUG:   [             142]     << Broadcast (F) -> TV (0): POLL
TRAFFIC: [             142]     << f0
DEBUG:   [             142]     usbcec: updating line timeout: 3
DEBUG:   [             204]     >> POLL sent
DEBUG:   [             204]     TV (0): device status changed into 'present'
DEBUG:   [             204]     << requesting vendor ID of 'TV' (0)
TRAFFIC: [             204]     << f0:8c
TRAFFIC: [             408]     >> 0f:87:00:00:f0
DEBUG:   [             408]     TV (0): vendor = Samsung (0000f0)
DEBUG:   [             408]     expected response received (87: device vendor id)
DEBUG:   [             408]     replacing the command handler for device 'TV' (0)
DEBUG:   [             408]     registering new CEC client - v6.0.2
DEBUG:   [             408]     >> TV (0) -> Broadcast (F): device vendor id (87)
DEBUG:   [             423]     usbcec: autonomous mode = disabled
DEBUG:   [             439]     usbcec: logical address = Recorder 1
DEBUG:   [             454]     usbcec: device type = recording device
DEBUG:   [             486]     usbcec: logical address mask = 206
TRAFFIC: [             501]     >> 40:8f
DEBUG:   [             564]     usbcec: physical address = 1000
DEBUG:   [             595]     usbcec: auto power on = disabled
TRAFFIC: [             595]     >> 04:90:01
DEBUG:   [             595]     CEC::CCECClient::SetClientVersion - using client version '6.0.2'
DEBUG:   [             595]     TV (0): power status changed from 'unknown' to 'standby'
NOTICE:  [             595]     setting HDMI port to 2 on device TV (0)
DEBUG:   [             595]     CEC::CCECClient::SetConfiguration: double tap timeout = 200ms, repeat rate = 0ms, release delay = 500ms
DEBUG:   [             595]     detecting logical address for type 'recording device'
DEBUG:   [             595]     trying logical address 'Recorder 1'
DEBUG:   [             595]     << Recorder 1 (1) -> Recorder 1 (1): POLL
TRAFFIC: [             595]     << 11
DEBUG:   [             642]     CEC transmission - received response - TRANSMIT_FAILED_ACK
TRAFFIC: [             642]     << 11
DEBUG:   [             689]     CEC transmission - received response - TRANSMIT_FAILED_ACK
DEBUG:   [             689]     >> POLL not sent
DEBUG:   [             689]     using logical address 'Recorder 1'
DEBUG:   [             689]     Recorder 1 (1): device status changed into 'handled by libCEC'
DEBUG:   [             689]     Recorder 1 (1): power status changed from 'unknown' to 'on'
DEBUG:   [             689]     Recorder 1 (1): vendor = Pulse Eight (001582)
DEBUG:   [             689]     Recorder 1 (1): CEC version 1.4
DEBUG:   [             689]     CEC::CCECClient::AllocateLogicalAddresses - device '0', type 'recording device', LA '1'
DEBUG:   [             689]     usbcec: updating ackmask: 0002
DEBUG:   [             704]     Recorder 1 (1): osd name set to 'CECTester'
DEBUG:   [             704]     Recorder 1 (1): menu language set to 'eng'
DEBUG:   [             704]     CEC::CUSBCECAdapterCommunication::GetPhysicalAddress - trying to get the physical address via ADL
DEBUG:   [             705]     CEC::CUSBCECAdapterCommunication::GetPhysicalAddress - ADL returned physical address 0000
DEBUG:   [             705]     CEC::CUSBCECAdapterCommunication::GetPhysicalAddress - trying to get the physical address via nvidia driver
DEBUG:   [             705]     CEC::CUSBCECAdapterCommunication::GetPhysicalAddress - nvidia driver returned physical address 0000
DEBUG:   [             705]     CEC::CUSBCECAdapterCommunication::GetPhysicalAddress - trying to get the physical address from the OS
DEBUG:   [             706]     CEC::CUSBCECAdapterCommunication::GetPhysicalAddress - OS returned physical address 0000
NOTICE:  [             706]     setting HDMI port to 2 on device TV (0)
DEBUG:   [             706]     Recorder 1 (1): physical address changed from ffff to 2000
DEBUG:   [             706]     << Recorder 1 (1) -> broadcast (F): physical address 2000
TRAFFIC: [             706]     << 1f:84:20:00:01
DEBUG:   [             861]     usbcec: updated physical address: 1000 -> 2000
DEBUG:   [             861]     using device/input physical address 0000
NOTICE:  [             861]     CEC client registered: libCEC version = 6.0.2, client version = 6.0.2, firmware version = 12, firmware build date: Tue Apr 28 14:20:49 2020 +0000, logical address(es) = Recorder 1 (1) , base device: TV (0), HDMI port number: 2, physical address: 2.0.0.0, compiled using MSVC 19.26.28806.0 (x86), features: P8_USB, P8_detect
DEBUG:   [             861]     << Recorder 1 (1) -> TV (0): OSD name 'CECTester'
TRAFFIC: [             861]     << 10:47:43:45:43:54:65:73:74:65:72
DEBUG:   [             876]     usbcec: eeprom updated
DEBUG:   [            1142]     << requesting power status of 'TV' (0)
TRAFFIC: [            1142]     << 10:8f
waiting for input
DEBUG:   [            1204]     expected response received (90: report power status)
TRAFFIC: [            1298]     >> 01:90:01
DEBUG:   [            1299]     >> TV (0) -> Recorder 1 (1): report power status (90)
TRAFFIC: [            7979]     >> 40:8f
TRAFFIC: [            8073]     >> 04:90:01
TRAFFIC: [           16005]     >> 40:8f
TRAFFIC: [           16099]     >> 04:90:01
TRAFFIC: [           24027]     >> 40:8f
TRAFFIC: [           24121]     >> 04:90:01
as
DEBUG:   [           26778]     making Recorder 1 (1) the active source
DEBUG:   [           26778]     TV (0): power status changed from 'standby' to 'in transition from standby to on'
DEBUG:   [           26778]     usbcec: updating active source status: active
NOTICE:  [           26780]     >> source activated: Recorder 1 (1)
DEBUG:   [           26780]     sending active source message for 'Recorder 1'
NOTICE:  [           26780]     << powering on 'TV' (0)
TRAFFIC: [           26780]     << 10:04
NOTICE:  [           26843]     << Recorder 1 (1) -> broadcast (F): active source (2000)
TRAFFIC: [           26843]     << 1f:82:20:00
DEBUG:   [           26952]     << Recorder 1 (1) -> TV (0): menu state 'activated'
TRAFFIC: [           26952]     << 10:8e:00
TRAFFIC: [           32040]     >> 40:8f
TRAFFIC: [           32134]     >> 04:90:01
DEBUG:   [           32134]     TV (0): power status changed from 'in transition from standby to on' to 'standby'
TRAFFIC: [           35102]     >> 01:1a:01
DEBUG:   [           35102]     << Recorder 1 (1) -> TV (0): deck status 'stop'
TRAFFIC: [           35102]     << 10:1b:1a
DEBUG:   [           35104]     >> TV (0) -> Recorder 1 (1): give deck status (1A)
DEBUG:   [           37868]     << POLL: TV (0) -> Audio (5)
DEBUG:   [           37914]     << POLL: TV (0) -> Audio (5)
TRAFFIC: [           40039]     >> 0f:32:65:6e:67
DEBUG:   [           40039]     TV (0): power status changed from 'standby' to 'on'
DEBUG:   [           40039]     TV (0): menu language set to 'eng'
DEBUG:   [           40042]     >> TV (0) -> Broadcast (F): set menu language (32)
TRAFFIC: [           40117]     >> 40:8f
TRAFFIC: [           40211]     >> 04:90:00
TRAFFIC: [           40321]     >> 40:8f
TRAFFIC: [           40414]     >> 04:90:00
TRAFFIC: [           40492]     >> 40:9f
TRAFFIC: [           40571]     >> 01:8c
DEBUG:   [           40571]     << Recorder 1 (1) -> TV (0): vendor id Pulse Eight (1582)
TRAFFIC: [           40571]     << 1f:87:00:15:82
DEBUG:   [           40572]     >> TV (0) -> Recorder 1 (1): give device vendor id (8C)
TRAFFIC: [           40774]     >> 01:8f
DEBUG:   [           40774]     << Recorder 1 (1) -> TV (0): on
TRAFFIC: [           40774]     << 10:90:00
DEBUG:   [           40775]     >> TV (0) -> Recorder 1 (1): give device power status (8F)
TRAFFIC: [           40961]     >> 04:9e:05
DEBUG:   [           40961]     TV (0): CEC version 1.4
TRAFFIC: [           41024]     >> 01:46
DEBUG:   [           41024]     << Recorder 1 (1) -> TV (0): OSD name 'CECTester'
TRAFFIC: [           41024]     << 10:47:43:45:43:54:65:73:74:65:72
DEBUG:   [           41025]     >> TV (0) -> Recorder 1 (1): give osd name (46)
TRAFFIC: [           41477]     >> 01:a0:00:00:f0:23
TRAFFIC: [           41477]     << 10:a0:00:00:f0:24:00:80
DEBUG:   [           41479]     >> TV (0) -> Recorder 1 (1): vendor command with id (A0)
TRAFFIC: [           41758]     >> 01:9f
DEBUG:   [           41758]     << Recorder 1 (1) -> TV (0): cec version 1.4
TRAFFIC: [           41758]     << 10:9e:05
DEBUG:   [           41761]     >> TV (0) -> Recorder 1 (1): get cec version (9F)
TRAFFIC: [           41914]     >> 04:8c
TRAFFIC: [           41992]     >> 04:8f
TRAFFIC: [           42133]     >> 4f:87:00:1a:11
DEBUG:   [           42133]     Playback 1 (4): vendor = Google (001a11)
DEBUG:   [           42133]     device Playback 1 (4) status changed to present after command device vendor id
DEBUG:   [           42133]     >> Playback 1 (4) -> Broadcast (F): device vendor id (87)
TRAFFIC: [           42211]     >> 04:83
TRAFFIC: [           42320]     >> 40:90:00
DEBUG:   [           42320]     Playback 1 (4): power status changed from 'unknown' to 'on'
TRAFFIC: [           42461]     >> 4f:84:20:00:04
DEBUG:   [           42461]     Playback 1 (4): physical address changed from ffff to 2000
DEBUG:   [           42461]     resetting HDMI port and base device to defaults
NOTICE:  [           42461]     setting HDMI port to 1 on device TV (0)
DEBUG:   [           42461]     marking Recorder 1 (1) as inactive source
DEBUG:   [           42461]     usbcec: updating active source status: inactive
DEBUG:   [           42463]     >> Playback 1 (4) -> Broadcast (F): report physical address (84)
NOTICE:  [           42477]     >> source deactivated: Recorder 1 (1)
DEBUG:   [           42477]     Recorder 1 (1): physical address changed from 2000 to 1000
DEBUG:   [           42477]     << Recorder 1 (1) -> broadcast (F): physical address 1000
TRAFFIC: [           42477]     << 1f:84:10:00:01
DEBUG:   [           42664]     making Recorder 1 (1) the active source
DEBUG:   [           42664]     usbcec: updating active source status: active
NOTICE:  [           42680]     >> source activated: Recorder 1 (1)
DEBUG:   [           42680]     sending active source message for 'Recorder 1'
NOTICE:  [           42680]     << powering on 'TV' (0)
TRAFFIC: [           42680]     << 10:04
NOTICE:  [           42742]     << Recorder 1 (1) -> broadcast (F): active source (1000)
TRAFFIC: [           42742]     << 1f:82:10:00
DEBUG:   [           42852]     << Recorder 1 (1) -> TV (0): menu state 'activated'
TRAFFIC: [           42852]     << 10:8e:00
DEBUG:   [           42947]     usbcec: updated physical address: 2000 -> 1000
TRAFFIC: [           42947]     >> 04:46
DEBUG:   [           42963]     usbcec: eeprom updated
```

opdenkamp commented:
what you're trying to do here is not supported by the CEC spec indeed. you cannot make another device active source in CEC 1.4b.
you're instructing libCEC to take the same physical address as another device with these commands, and libCEC correctly detects the colliding addresses and resets its own, because it must be invalid.

you can send a raw active source command with that address in (`4f:82:20:00` in this specific setup) and the TV will accept that because it can't tell which device is actually sending the command. but libCEC only officially supports what the spec supports, which means that it only supports making itself active source in high level commands.

akirchhoff1 commented:
@opdenkamp I see. I guess I'll just have to live with being out of spec as what I'm doing works for my needs. You're welcome to close the pull request then.

opdenkamp commented:
the problem that I've got with this change is that it's a very confusing api change for something that shouldn't happen and isn't needed really.
you shouldn't be using the `-p 2` option here and then call `tx 4f:82:20:00`, which makes the adapter send the same command as that Google device would send to make itself active.
